File: leetcode/dp/l322.js
function coinChange(coins, amount) {
    let dp = new Array(amount + 1).fill(amount + 1);
    dp[0] = 0;
    for (let target = 1; target <= amount; target++) {
        for (let coin of coins) {
            if (target - coin < 0) continue;
            dp[target] = Math.min(dp[target], dp[target - coin] + 1);
        }
    }
    return dp[amount] == amount + 1 ? -1 : dp[amount];
};

File: src/l322.cpp
#include <algorithm>
#include <vector>
#include <cstdint>

class Solution {
public:
    // for loop
    // dp[amount] = std::min(dp[amount - coin1],... dp[amount - coinN]) + 1;
    // dp[amount] = dp[amount - coin] + 1;
    // 
    int coinChange(std::vector<int>& coins, int amount) {
        std::sort(coins.begin(), coins.end());
        std::vector<int> dp(amount + 1, -1); // -1 
        dp[0] = 0;
        for (int i = 1; i <= amount;i++) {
            for (auto coin : coins) {
                if (coin > i) break;
                if (dp[i - coin] == -1) continue;
                if (dp[i] == -1) { // 未初始化
                    dp[i] = dp[i - coin] + 1;
                } else { // 有数据，做对比
                    dp[i] = std::min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[amount];
    }
};